Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,527 @@
+  + Use of `SVE<T>` may be preferred if SVE will not be used for `FEATURE_SIMD`
+
+Given lack of available hardware and a lack of thorough understanding of the specification:
+
++ SVE will require a separate design
++ SVE is considered out of scope for this document
+
+## Miscellaneous
+### Handling Instruction Deprecation
+
+Deprecation of instructions should be relatively rare
+
++ Do not introduce an intrinsic for a feature that is currently deprecated
++ In event an assembly instruction is deprecated
+  1. Prefer emulation using alternate instructions if practical
+  2. Add `SetThrowOnDeprecated()` interface to allow developers to find these issues
+
+## Approved APIs
+
+The following sections document APIs which have completed the API review process.
+
+Until each API is approved it shall be marked "TBD Not Approved"
+
+### `All`
+
+TBD Not approved
+
+### `Aes`
+
+TBD Not approved
+
+### `Atomics`
+
+TBD Not approved
+
+### `Crc32`
+
+TBD Not approved
+
+### `Dcpop`
+
+TBD Not approved
+
+### `Dp`
+
+TBD Not approved
+
+### `Fcma`
+
+TBD Not approved
+
+### `Fp`
+
+TBD Not approved
+
+### `Fp16`
+
+TBD Not approved
+
+### `Jscvt`
+
+TBD Not approved
+
+### `Lrcpc`
+
+TBD Not approved
+
+### `Pmull`
+
+TBD Not approved
+
+### `Ras`
+
+TBD Not approved
+
+### `Sha1`
+
+TBD Not approved
+
+### `Sha2`
+
+TBD Not approved
+
+### `Sha3`
+
+TBD Not approved
+
+### `Sha512`
+
+TBD Not approved
+
+### `Simd`
+
+```C#
+namespace System.Runtime.Intrinsics.Arm.Arm64
+{
+    /// <summary>
+    /// This class provides access to the Arm64 AdvSIMD intrinsics
+    ///
+    /// Arm64 CPU indicate support for this feature by setting
+    /// ID_AA64PFR0_EL1.AdvSIMD == 0 or better.
+    /// </summary>
+    [CLSCompliant(false)]
+    public static class Simd
+    {
+        /// <summary>
+        /// IsSupported property indicates whether any method provided 
+        /// by this class is supported by the current runtime.
+        /// </summary>
+        public static bool IsSupported { get; }
+
+        /// <summary>
+        /// Vector absolute value
+        /// Corresponds to vector forms of ARM64 ABS & FABS
+        /// </summary>
+        public static Vector64<byte>    Abs(Vector64<sbyte>   value);
+        public static Vector64<ushort>  Abs(Vector64<short>   value);
+        public static Vector64<uint>    Abs(Vector64<int>     value);
+        public static Vector64<ulong>   Abs(Vector64<long>    value);
+        public static Vector64<float>   Abs(Vector64<float>   value);
+        public static Vector128<byte>   Abs(Vector128<sbyte>  value);
+        public static Vector128<ushort> Abs(Vector128<short>  value);
+        public static Vector128<uint>   Abs(Vector128<int>    value);
+        public static Vector128<ulong>  Abs(Vector128<long>   value);
+        public static Vector128<float>  Abs(Vector128<float>  value);
+        public static Vector128<double> Abs(Vector128<double> value);
+
```

Comment:
Choice between Abs<T>(Vector128<T> value) and range of overloads is a tradoff: with generics only available constraint is `struct` which allows instantiation of Vectors with DateTime, Brotli Decoder and Encoder etc what may be limited by analyzers or users will face runtime exceptions, with overloads there are just more methods but calling them is perfectly type safe. I prefer the second choice, however, currently Intel intrinsics are partially implemented with egnerics

Reply:
I think the initial Intel intrinsics were small enough that this wasn't an issue. However, later intrinsics (such as the AVX-512 family) may have the same issue.

Maybe this is worth revisiting WRT both Intel and Arm in order to help prevent an API explosion.